Project/combat.py

```diff
@@ -0,0 +1,69 @@
+import random
+from player import Player
+
+class Combat:
+    """Handles turn-based combat mechanics."""
+    
+    def __init__(self, player):
+        self.player = player
+        self.enemy = self.generate_enemy()
+
+    def generate_enemy(self):
+        """Dynamically generates an enemy with randomized stats."""
+        enemies = [
+            {"name": "Goblin", "health": 50, "attack": 10, "defense": 2, "evasion": 5},
+            {"name": "Orc", "health": 80, "attack": 15, "defense": 4, "evasion": 3},
+            {"name": "Dark Mage", "health": 60, "attack": 12, "defense": 3, "evasion": 7}
+        ]
+        return random.choice(enemies)
+
+    def calculate_damage(self, attacker_strength, enemy_defense):
+        """Uses expressions to calculate attack damage dynamically."""
+        base_damage = attacker_strength * random.uniform(1.2, 1.8)  # Adds some randomness
+        final_damage = max(base_damage - enemy_defense, 1)  # Ensure at least 1 damage
+        return round(final_damage, 1)
+
+    def player_attack(self):
+        """Player's turn to attack."""
+        if random.randint(1, 100) <= self.enemy["evasion"]:  # Enemy dodges attack
+            print(f"💨 {self.enemy['name']} dodged the attack!")
+            return
+        
+        damage = self.calculate_damage(self.player.stats["strength"], self.enemy["defense"])
+        self.enemy["health"] -= damage
+        print(f"🗡 {self.player.name} attacks! {self.enemy['name']} loses {damage} HP.")
+
+        # Special Abilities
+        if self.player.role == "Warrior" and random.randint(1, 100) <= 20:
+            print(f"💥 {self.player.name} lands a **CRITICAL STRIKE** for double damage!")
+            self.enemy["health"] -= damage  # Apply extra hit
+
+        if self.player.role == "Mage" and self.player.mana >= 20:
+            print(f"🔥 {self.player.name} casts **Fireball** on {self.enemy['name']}!")
+            self.enemy["health"] -= 15
+            self.player.mana -= 20  # Spend mana
+
+    def enemy_attack(self):
+        """Enemy's turn to attack."""
+        if random.randint(1, 100) <= self.player.stats.get("agility", 0) * 2:  # Higher agility = better dodge chance
```

Comment:
If agility is 0 or missing, the enemy attack would always be successful, consider a default value e.g: 1 to ensure fairness

```suggestion
        if random.randint(1, 100) <= self.player.stats.get("agility", 0) * 2:  # Higher agility = better dodge chance
```